packages/dendron-plugin-views/src/hooks/useGraphElements.tsx

```diff
@@ -34,166 +35,333 @@ const DEFAULT_CLASSES = ""; //graph-view
+    const isActive = activeNote && note.id === activeNote.id;
     nodes.push({
       data: {
-        id: parentNote.id,
-        label: parentNote.title,
+        id: note.id,
+        label: note.title,
         group: "nodes",
-        fname: parentNote.fname,
-        color: getNoteColor({ fname: parentNote.fname, notes }),
-        stub: _.isUndefined(parentNote.stub) ? false : parentNote.stub,
+        fname: note.fname,
+        color: getNoteColor({ fname: note.fname, notes }),
+        stub: _.isUndefined(note.stub) ? false : note.stub,
+        localRoot: distance === 0,
       },
-      classes: `${DEFAULT_NODE_CLASSES} parent ${getVaultClass(
-        parentNote.vault
-      )}`,
+      classes,
+      selected: isActive,
     });
-  }
 
-  // Add connected nodes
-  nodes.push(
-    ...Object.values(connectedNotes).map((note) => {
-      return {
-        data: {
-          id: note.id,
-          label: note.title,
-          group: "nodes",
-          fname: note.fname,
-          color: getNoteColor({ fname: note.fname, notes }),
-          stub: _.isUndefined(note.stub) ? false : note.stub,
-        },
-        classes: `${DEFAULT_NODE_CLASSES} ${getVaultClass(note.vault)}`,
-      };
-    })
-  );
+    if (data.distance < maxDistance) {
+      const noteVaultClass = getVaultClass(note.vault);
+
+      //Add parents and grandparents. Distance 0 indicates the immediate parent, isParent indicate the ancestor
+      const parentNote =
+        (data.distance === 0 || data.isParent) && note.parent
+          ? notes[note.parent]
+          : undefined;
+      if (parentNote) {
+        nodesQueue.enqueue({
+          note: parentNote,
+          distance: data.distance + 1,
+          isParent: true,
+          classes: `${DEFAULT_NODE_CLASSES} parent ${getVaultClass(
+            parentNote.vault
+          )}`,
+        });
+        edges.hierarchy.push({
+          data: {
+            group: "edges",
+            id: `${parentNote.id}_${note.id}`,
+            source: parentNote.id,
+            target: note.id,
+            fname: parentNote.fname,
+            stub: _.isUndefined(parentNote.stub) ? false : parentNote.stub,
+          },
+          classes: `${DEFAULT_EDGE_CLASSES} hierarchy ${noteVaultClass}`,
+        });
+      }
+      // do not include children of parent/grandparent
+      const children = data.isParent ? [] : note.children;
 
+      // Setup the edges for children now:
+      children.forEach((child: string) => {
+        const childNote = notes[child];
+        if (childNote) {
+          edges.hierarchy.push({
+            data: {
+              group: "edges",
+              id: `${note.id}_${childNote.id}`,
+              source: note.id,
+              target: childNote.id,
+              fname: note.fname,
+              stub: false,
+            },
+            classes: `${DEFAULT_EDGE_CLASSES} hierarchy ${noteVaultClass}`,
+          });
+
+          nodesQueue.enqueue({
```

Comment:
In this case, won't we iterate over the active note multiple times, since active note === active note's child's parent?

Reply:
I don't think so, this is because we add a property `isParent: true` for active note's parent and only enqueue parent if it is either active note's parent or its ancestor. Parent of child notes are not added to the queue.